### ISSUE TITLE
Closures accepting structs with lifetimes

### ISSUE THREAD

**Issue by Michael-F-Bryan** [getditto/safer_ffi#69]
I ran into an annoying issue with `BoxDynFnMut1` and friends when you try to pass around a struct with a non-`'static` lifetime due to the lack of higher-kinded types.

Basically, I want to let users set their own logging function. In normal Rust you'd write something like this:

```rust
pub fn set_logger(log_func: Box<dyn Fn(&log::Record<'_>)>) { ... }
```

The equivalent using `safer_ffi` would be (in theory) something like this:

```rust
#[ffi_export]
pub fn set_logger(log_func: for<'a> BoxDynFnMut1<(), LogRecord<'a>>) { ... }

#[derive_ReprC]
#[repr(C)]
pub struct LogRecord<'a> {
    level: LogLevel,
    target: str_ref<'a>,
}
```

Obviously `for<'a> SomeStruct` isn't valid syntax and you get an *"Error: expected trait, found struct `BoxDynFnMut1`"* error, however without the `for<'a>` you get *"Error: cannot infer an appropriate lifetime due to conflicting requirements"* when you try to use it in practice.

This is a concrete example of the *"improve support for closures"* mentioned in #47.

**Comment by danielhenrymantilla** [comment]
Indeed! FWIW, the least bad solution for higher-ranked _closures_, right now, will be an API very similar to https://docs.rs/stackbox/0.1.2/stackbox/macro.custom_dyn.html . If you have suggestions ideas for these things, I'm all ears 🙂 

You can also test the `ditto` internal branch, which already supports higher-ranked _function pointers_ (stateless):

```rust
#[derive_ReprC]
#[repr(transparent)]
struct MyLoggerFnPointer(
    extern "C" fn(&log::Record<'_>),
);
```

**Comment by danielhenrymantilla** [comment]
I know this is an old issue, but I wanted to gauge the situation at the moment, @Michael-F-Bryan:

  - `safer-ffi` now supports FFI-safe `dyn` traits which can be `Box`ed, and which hopefully do support this generic lifetime param in input;
  - thanks to https://docs.rs/higher-kinded-types, it would be possible to express some <code>ForFn<[ForLt!]\(&log::Record<'_>)></code> kind of type, at least in the covariant case
      - (or <code>ForFn<[For!]\(<'a, 'b> = &'a log::Record<'b>)></code> in the invariant case)

[ForLt!]: https://docs.rs/higher-kinded-types/0.1.1/higher_kinded_types/macro.ForLt.html

[For!]: https://docs.rs/higher-kinded-types/0.1.1/higher_kinded_types/extra_arities/macro.For.html
